src/merkle/new_smt/mod.rs

```diff
@@ -0,0 +1,272 @@
+                            let old_value = kv_pairs[pos].1;
+                            kv_pairs[pos].1 = value;
+
+                            Some(old_value)
+                        }
+                        Err(pos) => {
+                            kv_pairs.insert(pos, (key, value));
+
+                            None
+                        }
+                    }
+                }
+            },
+            None => {
+                self.leaves.insert(leaf_index.value(), NewSmtLeaf::Single((key, value)));
+
+                Some(Self::Value::default())
+            }
+        }
+    }
+
+    fn get_leaf(&self, key: &Self::Key) -> Self::Leaf {
+        let leaf_pos = LeafIndex::<NEW_SMT_DEPTH>::from(*key).value();
+
+        match self.leaves.get(&leaf_pos) {
+            Some(leaf) => leaf.clone(),
+            None => NewSmtLeaf::Single((
+                *key,
+                Word::from(*EmptySubtreeRoots::entry(self.depth(), self.depth())),
+            )),
+        }
+    }
+
+    fn hash_leaf(leaf: &Self::Leaf) -> RpoDigest {
+        leaf.hash()
+    }
+}
+
+impl Default for NewSmt {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// KEY
+// ================================================================================================
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct NewSmtKey {
+    word: Word,
+}
+
+impl From<NewSmtKey> for LeafIndex<NEW_SMT_DEPTH> {
+    fn from(key: NewSmtKey) -> Self {
+        let most_significant_felt = key.word[0];
+        Self::new_max_depth(most_significant_felt.as_int())
+    }
+}
+
+impl Ord for NewSmtKey {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Note that the indices are reversed. This is because a `Word` is treated as little-endian
+        // (i.e. most significant Felt is at index 3). In comparing integer arrays, we want to
+        // compare the most significant felts first, and so on until the least signifcant felt.
+        let self_word: [u64; 4] = [
+            self.word[3].as_int(),
+            self.word[2].as_int(),
+            self.word[1].as_int(),
+            self.word[0].as_int(),
+        ];
+        let other_word: [u64; 4] = [
+            other.word[3].as_int(),
+            other.word[2].as_int(),
+            other.word[1].as_int(),
+            other.word[0].as_int(),
+        ];
+
+        self_word.cmp(&other_word)
+    }
+}
+
+impl PartialOrd for NewSmtKey {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+// LEAF
+// ================================================================================================
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub enum NewSmtLeaf {
+    Single((NewSmtKey, Word)),
+    Multiple(Vec<(NewSmtKey, Word)>),
+}
+
+impl NewSmtLeaf {
```

Comment:
We'd also need to implement something like `to_elements()` here - e.g.:
```Rust
pub fn to_elements(&self) -> Vec<Felt>

pub fn into_elements(self) -> Vec<Felt>
```
We will probably need these to inject the SMT into the advice provider.

Reply:
Any reason why we shouldn't implement `From` for that, instead of custom methods?

Reply:
No strong reason. Sometimes it make the code more readable, but in this case probably doesn't make a difference.